### ISSUE TITLE
In the facet box display the facet expression used 

### ISSUE THREAD

**Issue by magdmartin** [OpenRefine/OpenRefine#745]
When you combine multiple true/false facet on the same column it can be hard to remember what the action behind each one (isblank or a custom facet), see image attached

![https mail-attachment googleusercontent com attachment u 0 ui 2 ik 39313f2a28 view att th 13f520c0a725495e attid 0 1 disp inl](https://f.cloud.github.com/assets/2589584/674553/dbc4b606-d8db-11e2-8086-3e01bc950699.png)

Currently the work around is to click change, close the transform windows without doing anything, to display the GREL expression in the facet box.

This could be expended directly or display in a different manner

**Comment by thadguidry** [comment]
I think the following should happen:
1. when a user clicks change, the `expressionDiv` displays.
2. once the user clicks anywhere in the `expressionDiv` then the Expression editor dialog appears (rather than the current auto launching of the Expression editor dialog when the user clicks the `change` link)

So the fix is to just remove that auto launching behavior, because the `expressionDiv` clicking functionality already exists and works.
So this is really a good first issue for someone.

**Comment by thadguidry** [comment]
I think a better way to handle @magdmartin original need is:

1. add a new preference so that Facets will always show the expression in a 1 (current) or 2 (proposed) line expansion.  And there optionally could be a new control button **Show Exp** for that preference added to the Facet panel controls and where buttons slightly restyled, shifted, fonts reduced, etc. :
 
![image](https://user-images.githubusercontent.com/986438/85207467-8ade3380-b2ee-11ea-94cc-86ced7899163.png)

2. Additionally, it would be nice to allow the `left-panel` to be able to be resized horizontally by grabbing the divider bar to allow the user to have room to see the expressions just a bit more:

![image](https://user-images.githubusercontent.com/986438/85207643-a1d15580-b2ef-11ea-84f6-4ec328ead2c5.png)

**Comment by man-See** [comment]
Hi @wetneb , I am Mansi. I was getting familiar with OpenRefine. 

Here, are some screenshots.

![Slice 1 (12)](https://user-images.githubusercontent.com/119604428/223719202-37ca4ad9-e7ea-481f-8516-5616cd3153a6.png)

In this, I made a 'details' link instead of 'change'. When I was using OpenRefine I click on change to see the expression -> then a window pops up -> I cancelled it -> then I hover over the expression to edit. I think there are some extra steps to make a change. So, I come up with a solution for it.

![Slice 2 (10)](https://user-images.githubusercontent.com/119604428/223720527-ddf93ed3-fe2c-4ca4-85c2-c2de8fb23d1b.png)

When I click on the details, a window will pop up and have an option for editing. I also wanted to add a custom facet option but didn't feel like I have a proper understanding of the same.

You can also check this Figma file for the same
https://www.figma.com/file/UuKfkDYg7n0EF3lJ2J0Uzj/Untitled?node-id=0-1&t=oxOWDJ40G6VrJgf8-0

**Comment by wetneb** [comment]
I think this does not address the need expressed in the issue. We already have a "change" link in the facet header which does essentially the same thing as your "details" link.

**Comment by maaannsiP** [comment]
I admit that it did not address the need. I was thinking about it, from the day I posted that comment. I agree that details and change did the same thing. So, that was not needed to just change the name of the link. The problem stated above is when they try to combine true/false facets in the same column there is no way to know from which action they did it. 

One solution is to add an action tab in the change link window.
The other is to show by hover.

Any feedback on this?

**Comment by wetneb** [comment]
Note that one way to address the original problem @magdmartin had, without displaying the GREL expression (which could be a bit scary for users who are not familiar with that) would be to change the facet so that it does not return "true" or "false" but more informative values which hint at the criterion used.

For instance this is an approach I have used when implementing the faceting for issues from the Wikibase extension:
![image](https://user-images.githubusercontent.com/309908/177998588-5a1ff5e9-b529-4025-bd5b-8d0ecb8011e2.png)

The same principle could be applied to facets such as "facet by blank": instead of returning "true" or "false", they would show "blank" or "not blank".

Personally I find it annoying to have to go through the thought process of "I want to select the non-blank cells, so because this facet is about blank cells, I should pick the option `false`." It's easy to mess things up and associate `true` with "some content is present" and `false` with "some content is absent". So this proposal would also help reduce that cognitive load (at least for me).

It's also quite easy to implement: it's just about changing the GREL expression defining the facet from `value.isBlank()` to `if(value.isBlank(), "blank", "not blank")`.

Of course we could think of a lot of other designs (not necessarily based on the text facet).

One slight concern I have about my proposal is how to localize those strings. Because the strings would be included in the GREL expressions defining the facet, those strings would also get included in the project history, in the original language they were done with. So if someone else re-applies a list of operations made in another language, the facet values could make little sense to them. But we already have localized operation descriptions in JSON histories so it would not be a completely new problem.

**Comment by ostephens** [comment]
I'm 100% in favour of changing the labelling in an 'is Blank' facet to be more explicit - having to explain to people "true means "it is true that it is blank" is something that comes up in every single training session I do!

This would apply to several of our "customised facets" (in the Facet -> Customised facets drop down menu):

* Facet by blank (null or empty string)
* Facet by error
* Facet by null
* Facet by empty string
* Duplicates

Another option for handling the blanks/nulls/empties in a more intuitive way might be to actually combine these facets into a single customised facet so the user doesn't have to decide up front which they are looking for - although might be worth having some user consultation on that idea my immediate thought is that this would be a good approach

This won't resolve the situation where you have multiple different custom expressions and can't remember which is which - but I'd be in favour of being able to give the facet a name/label to handle that rather than exposing the GREL all the time (which would be v distracting even for those of us comfortable writing it)

**Comment by tfmorris** [comment]
Providing a facet labelling facility with a default, translatable, label like "Is Blank?" is a much more easily localized solution than changing the GREL expressions.

**Comment by ostephens** [comment]
This is a really good point @tfmorris - I'd badly overlooked the localisation aspect.

**Comment by wetneb** [comment]
As far as I can tell, only changing the facet name (in a localized way) has the same "issue", since that name will also end up in the facet configuration (and therefore JSON workflow).

But it's a really minor issue since the JSON workflow is not really human-readable anyway and facets are not persisted.
So to me the benefits of changing the GREL expression, including localized strings for both options, seems to be a net win. I am not sure where you see a difference @tfmorris?

**Comment by tfmorris** [comment]
I don't have time right now to investigate, so I'll take your word for it that they're equivalent. I guess my naive view was that a label less critical than the GREL code. Does this mean that our undo/redo histories break if the user changes their interface language?

**Comment by wetneb** [comment]
> Does this mean that our undo/redo histories break if the user changes their interface language?

No, it would only mean that reading the JSON representation of the history is perhaps less easy if you are not familiar with the language it was originally made in.
